<a href="https://colab.research.google.com/github/alupo933/Clustering-Work-for-Machine-Learning/blob/main/Final_PCA_Exercise_(CORE).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Task
The task is to perform PCA to speed up a classification algorithm on a high-dimensional dataset.

We will use the MNIST digits dataset, which comes pre-installed in sklearn. This dataset has 8x8 pixel images of handwritten digits 0-9. Your task is to classify these to determine which digits they are.

Use PCA to lower the dimensions in this dataset while retaining 95% of the variance.  You can do this when instantiating the PCA by giving the `n_components=` argument a float between 0 and 1.

In [16]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.preprocessing import LabelEncoder, StandardScaler
from sklearn.decomposition import PCA
from sklearn.pipeline import make_pipeline
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score
from sklearn.model_selection import train_test_split

In [17]:
# Load the dataset
from sklearn.datasets import fetch_openml
mnist = fetch_openml('mnist_784')

# View the shape of the dataset
mnist.data.shape

(70000, 784)

The dataset has shape (70000, 784), meaning we are working with 70,000 images with 784 dimensions!

Note that you can access the X data using `mnist.data` and access the target using `mnist.target`.

If you get an error using the above code, you can also load the data using:

In [18]:
# from keras.datasets import mnist

# load the the data
# (X_train, y_train), (X_test, y_test) = mnist.load_data()

# reshape the data
# X_train = X_train.reshape(X_train.shape[0], -1)
# X_test = X_test.reshape(X_test.shape[0], -1)

In [19]:
# First I'll save the data into features and a target
X = mnist.data
y = mnist.target

In [20]:
X.head()

pixel1  pixel2  pixel3  pixel4  ...  pixel781  pixel782  pixel783  pixel784
0     0.0     0.0     0.0     0.0  ...       0.0       0.0       0.0       0.0
1     0.0     0.0     0.0     0.0  ...       0.0       0.0       0.0       0.0
2     0.0     0.0     0.0     0.0  ...       0.0       0.0       0.0       0.0
3     0.0     0.0     0.0     0.0  ...       0.0       0.0       0.0       0.0
4     0.0     0.0     0.0     0.0  ...       0.0       0.0       0.0       0.0

[5 rows x 784 columns]

In [21]:
# Look for missing values
X.isnull().sum()

pixel1      0
pixel2      0
pixel3      0
pixel4      0
pixel5      0
           ..
pixel780    0
pixel781    0
pixel782    0
pixel783    0
pixel784    0
Length: 784, dtype: int64

There is no missing data

In [22]:
X.duplicated().sum()


0

There are no duplicates

In [23]:
# Look at the class balance
y.value_counts()

1    7877
7    7293
3    7141
2    6990
9    6958
0    6903
6    6876
8    6825
4    6824
5    6313
Name: class, dtype: int64

In [24]:
X.shape
# we can see this matches the original  DataFrame

(70000, 784)

In [25]:
# Instantiate and fit/transform the standard scaler
scaler = StandardScaler()

scaled_df = scaler.fit_transform(X)

Next I will Instantiate and fir the data usiong PCA

In [26]:
pca = PCA()
pca.fit(scaled_df)

PCA()

In [27]:
pca95 = PCA(n_components=.95)

In [28]:
#Perform a test train split
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state = 42)

In [29]:
# Next we'll make PCA Pipeline with a Standard Scaler
transformer95 = make_pipeline(scaler, pca95)

In [39]:
# Instantiate the logistic Regression model
logreg = LogisticRegression(max_iter=900)

In [40]:
# create a model pipeline for the 95% explained variance
logpipe95 = make_pipeline(transformer95, logreg)

Fit the Logistic Regression pipelines to the training data

In [41]:
logpipe95.fit(X_train, y_train)

Pipeline(steps=[('pipeline',
                 Pipeline(steps=[('standardscaler', StandardScaler()),
                                 ('pca', PCA(n_components=0.95))])),
                ('logisticregression', LogisticRegression(max_iter=900))])

# Score the Models

In [38]:
print('---'*10)
print('95% Explained Variance')
print('---'*10)
print('Training Accuracy', logpipe95.score(X_train, y_train))
print('Testing Accuracy', logpipe95.score(X_test, y_test))

------------------------------
95% Explained Variance
------------------------------
Training Accuracy 0.9365523809523809
Testing Accuracy 0.9217714285714286
